cmd/gcp-controller-manager/readonly_resource_csr_approver.go

```diff
@@ -0,0 +1,310 @@
+		{
+			name:          "rbac validator",
+			authFlowLabel: "readonly_resource_rbac_exist",
+			validate:      validateRbac,
+		},
+	}
+}
+
+func validateCSRUsage(content readonlyResourceCSRRequestContent) readonlyResourceCSRRequestResult {
+	csr := content.csr
+	if csr == nil || len(csr.Spec.Usages) == 0 {
+		klog.Errorf("csr usage is empty")
+		return readonlyResourceCSRRequestResult{
+			result:  false,
+			err:     nil,
+			message: "csr usage is empty",
+		}
+	}
+
+	usageMap := map[capi.KeyUsage]struct{}{}
+	for _, u := range kubeletClientUsages {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, usage := range csr.Spec.Usages {
+		if _, ok := usageMap[usage]; !ok {
+			klog.Errorf("csr usage %v is not allowed", usage)
+			return readonlyResourceCSRRequestResult{
+				result:  false,
+				err:     nil,
+				message: fmt.Sprintf("csr usage %v is not allowed", usage),
+			}
+		}
+	}
+	return readonlyResourceCSRRequestResult{
+		result:  true,
+		err:     nil,
+		message: "csr usage is valid",
+	}
+}
+
+func validatePodAnnotation(content readonlyResourceCSRRequestContent) readonlyResourceCSRRequestResult {
+	if content.csr == nil || len(content.csr.Spec.Extra) == 0 {
+		klog.Errorf("csr Extra is empty")
+		return readonlyResourceCSRRequestResult{
+			result:  false,
+			err:     nil,
+			message: "csr Extra is empty",
+		}
+	}
+
+	podNames, ok := content.csr.Spec.Extra[podNameKey]
+	if !ok {
+		klog.Errorf("csr does not have pod name attached")
+		return readonlyResourceCSRRequestResult{
+			result:  false,
+			err:     fmt.Errorf("csr does not have pod name attached"),
+			message: "csr does not have pod name attached",
+		}
+	}
+
+	for _, podName := range podNames {
+		pod, err := content.controllerContext.client.CoreV1().Pods("").Get(content.context, podName, metav1.GetOptions{})
+		if err != nil {
+			klog.Errorf("Error when get pod %s, error: %v", podName, err)
+			return readonlyResourceCSRRequestResult{
+				result:  false,
+				err:     fmt.Errorf("Error when get pod %s, error: %v", podName, err),
+				message: fmt.Sprintf("Error when get pod %s, error: %v", podName, err),
+			}
+		}
+
+		hasAnnotation := false
+		for _, annotation := range pod.Annotations {
+			if annotation == kubeletAPILimitedReaderAnnotationKey {
+				hasAnnotation = true
+			}
+		}
+		if !hasAnnotation {
+			klog.Errorf("Pod %s not have annotation: %s", podName, kubeletAPILimitedReaderAnnotationKey)
+			return readonlyResourceCSRRequestResult{
+				result:  false,
+				err:     nil,
+				message: fmt.Sprintf("Pod %s not have annotation: %s", podName, kubeletAPILimitedReaderAnnotationKey),
+			}
+		}
+		klog.Errorf("Pod %s have annotation: %s", podName, kubeletAPILimitedReaderAnnotationKey)
+	}
+
+	return readonlyResourceCSRRequestResult{
+		result:  true,
+		err:     nil,
+		message: fmt.Sprintf("Annotation validation pass"),
+	}
+}
+
+// validateRbac include SAR validation
+func validateRbac(content readonlyResourceCSRRequestContent) readonlyResourceCSRRequestResult {
```

Comment:
Its weird that we are performing authorization checks here.

Reply:
I just want to make all check to be portable